### PR TITLE
Persistent globally unique DB ID in manifest

### PR DESCRIPTION
**Pull request by vjnadimpalli**
Each DB has a globally unique ID. A DB can be physically copied around, or backed-up and restored, and the users should be identify the same DB. This unique ID right now is stored as plain text in file IDENTITY under the DB directory. This approach introduces at least two problems: (1) the file is not checksumed; (2) the source of truth of a DB is the manifest file, which can be copied separately from IDENTITY file, causing the DB ID to be wrong. 
The goal of this PR is solve this problem by moving the  DB ID to manifest. To begin with we will write to both identity file and manifest. Write to Manifest is controlled via the flag write_dbid_to_manifest in Options and default is false. 

Test Plan:
1. Added unit tests. 
2. Verified forward compatibility by checking out 6.2.2 and reading an example Manifest containing DB ID using ldb and it rightly ignored the DBID while printing out Manifest.
3. Make check with write_dbid_to_maifest = true

### CODE REVIEW

File: db/version_edit.cc [hunk 1]
```diff
@@ -31,6 +33,7 @@ enum Tag : uint32_t {
   // 8 was used for large value refs
   kPrevLogNumber = 9,
   kMinLogNumberToKeep = 10,
+  kDbId = 11,
```
[the remark this answers is not in the document]
Comment (vjnadimpalli) [reply db/version_edit.cc]: Accidental..will change it to ignoreable.

File: db/version_set.cc [hunk 2]
```diff
@@ -4131,61 +4128,54 @@ Status VersionSet::ApplyOneVersionEditToBuilder(
...
 
-Status VersionSet::ExtractInfoFromVersionEdit(
-    ColumnFamilyData* cfd, const VersionEdit& edit, bool* have_log_number,
-    uint64_t* log_number, bool* have_prev_log_number,
-    uint64_t* previous_log_number, bool* have_next_file, uint64_t* next_file,
-    bool* have_last_sequence, SequenceNumber* last_sequence,
-    uint64_t* min_log_number_to_keep, uint32_t* max_column_family) {
+Status VersionSet::ExtractInfoFromVersionEdit(ColumnFamilyData* cfd,
+                                              const VersionEdit& from_edit,
+                                              VersionEdit* to_edit) {
```
[the remark this answers is not in the document]
Comment (vjnadimpalli) [reply db/version_set.cc]: Fixing it using alias -> using VersionEditParams = VersionEdit.

File: db/db_basic_test.cc [hunk 3]
```diff
@@ -70,6 +70,39 @@ TEST_F(DBBasicTest, ReadOnlyDB) {
...
+
+  // Reopen and flush memtable.
+  Reopen(options);
+  Flush();
+  Close();
+  // Now check keys in read only mode.
+  ASSERT_OK(ReadOnlyReopen(options));
+  ASSERT_EQ("v3", Get("foo"));
+  ASSERT_EQ("v2", Get("bar"));
+  ASSERT_TRUE(db_->SyncWAL().IsNotSupported());
```
Comment (siying) [inline db/db_basic_test.cc]: The test is good. Can we also verify the DB ID returned is correct?

File: db/db_basic_test.cc [hunk 4]
```diff
@@ -424,12 +457,39 @@ TEST_F(DBBasicTest, ManifestRollOver) {
...
+    // id1 should match id2 because identity was not regenerated
+    ASSERT_EQ(id1.compare(id2), 0);
+
+    std::string idfilename = IdentityFileName(dbname_);
+    ASSERT_OK(env_->DeleteFile(idfilename));
+    Reopen(options);
+    std::string id3;
+    ASSERT_OK(db_->GetDbIdentity(id3));
+    if (options.write_dbid_to_manifest) {
+      ASSERT_EQ(id1.compare(id3), 0);
```
Comment (siying) [inline db/db_basic_test.cc]: nit: I believe you can use `ASSERT_EQ(id1, id3)`. In this case, when assertion fails, we also prints out the wrong value.

**Comment by riversand963** [comment]
Is it possible to add item 2 of test plan to tools/check_format_compatible.sh?

**Comment by socketpair** [comment]
Hi, guys. Just from sources:

```C
  // Historically DB ID has always been stored in Identity File in DB folder.
  // If this flag is true, the DB ID is written to Manifest file in addition
  // to the Identity file. By doing this 2 problems are solved
  // 1. We don't checksum the Identity file where as Manifest file is.
  // 2. Since the source of truth for DB is Manifest file DB ID will sit with
  //    the source of truth. Previously the Identity file could be copied
  //    independent of Manifest and that can result in wrong DB ID.
  // We recommend setting this flag to true.
  // Default: false
  bool write_dbid_to_manifest = false;
```

Why the recommended value is not set as default ?

**Pull request closed without merging**